### PR TITLE
Use comma separated multiple values for header

### PR DESCRIPTION
**Pull request by lilida**
Use set instead of add for the X-Auth-Request-Groups header.

## Description

When trying to test the latest group support https://github.com/oauth2-proxy/oauth2-proxy/pull/616, I found out the X-Auth-Request-Groups header has been truncated and only first header is returned. My environment is nginx-ingress 0.30.0. I tried to dig out why this happened. It turned out if we set the header multi-values directly instead add each value, it works fine. I suspect somewhere in the proxy code it calls header.Get() which only returns the first value in the header. But I am not able to find it out

## Motivation and Context
This is a bug needed to be fixed

## How Has This Been Tested?
Tested the headers are appended correctly for single and multiple group value

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by lilida** [comment]
cc: @stefansedich @NickMeves  This is one issue I have seen when tested the latest groups change

File: oauthproxy.go
Comment (NickMeves) [inline oauthproxy.go:979]: We would need to make this change to `X-Forwarded-Groups` as well.

@JoelSpeed I remember this PR used to have the groups joined explicitly into 1 string with `,` and you advised for `Add` instead.  Do you have more insight to the nuance of this and how these settings are called when the actual `ResponseWriter` is rendering these headers?

I see this distinction between `Get` vs `Values` on array headers, but not quickly finding how `ResponseWriter` renders:

```
// Get gets the first value associated with the given key.
// It is case insensitive; CanonicalMIMEHeaderKey is used
// to canonicalize the provided key.
// If there are no values associated with the key, Get returns "".
// To use non-canonical keys, access the map directly.
func (h MIMEHeader) Get(key string) string {
	if h == nil {
		return ""
	}
	v := h[CanonicalMIMEHeaderKey(key)]
	if len(v) == 0 {
		return ""
	}
	return v[0]
}

// Values returns all values associated with the given key.
// It is case insensitive; CanonicalMIMEHeaderKey is
// used to canonicalize the provided key. To use non-canonical
// keys, access the map directly.
// The returned slice is not a copy.
func (h MIMEHeader) Values(key string) []string {
	if h == nil {
		return nil
	}
	return h[CanonicalMIMEHeaderKey(key)]
}
```

Comment (NickMeves) [reply oauthproxy.go:979]: If only the first entry is ending up in the actual header, seems like the `http.ResponseWriter` is using `Get` where `X-Forwarded-Groups` that still had all the entries (but this was a `http.Request`) might be using `Values`...?

Comment (lilida) [reply oauthproxy.go:977]: Yeah. I only saw the first entry and  I have the same thought here. But I couldn't drill down to the exact line where Get is called. Also updated the X-Forwarded-Groups

**Comment by lilida** [comment]
@NickMeves @JoelSpeed Any idea on the issue and how we can proceed?

**Comment by JoelSpeed** [comment]
This is actually a bug in Nginx and not a bug in OAuth2 Proxy. If you load up your environment and take a look at the response headers from `/oauth2/auth` you will see that there is in fact multiple values for your groups under multiple headers

For example:
Capture

This is how the Go libraries set multi valued headers and it is in accordance with the [HTTP specification](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2).

> Multiple message-header fields with the same field-name MAY be present in a message if and only if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)]. It MUST be possible to combine the multiple header fields into one "field-name: field-value" pair, without changing the semantics of the message, by appending each subsequent field-value to the first, each separated by a comma. The order in which header fields with the same field-name are received is therefore significant to the interpretation of the combined field value, and thus a proxy MUST NOT change the order of these field values when a message is forwarded.

NGINX's problem is that when you call `auth_request_set`, it is seeing the three values and, where it should concatenate them and make them comma separated (`Group1,Group2,Everyone`), it just uses the first value.

I believe people have worked around this issue in the past using the NGINX Lua module but I'm not certain of that

**Comment by NickMeves** [comment]
So I actually didn't fully process the behavior of this was split headers (and therefore think we should look at `X-Forwarded-Groups` again as well if that has each group in a split header) -- reading that spec is a little confusing though, so feel free to disregard me if I'm understanding things incorrectly 😅 

I see the merits in slitting is this were called `X-Forwarded-Group` (singular).  But naming it `Groups` I think this should be consolidated to one header with a separator. (I dug up this old discussion): https://github.com/oauth2-proxy/oauth2-proxy/pull/616#discussion_r441064196

Even though it is less clean per HTTP spec, it fits better what enterprise upstream apps expect for SSO roles:

Here's an an example of Jenkins wanting that scheme for SSO groups: https://learn.akamai.com/en-us/webhelp/enterprise-application-access/enterprise-application-access/GUID-C5FBC2F3-EEC4-4C43-82BC-94E4B72B051F.html

And a sample from Graylog SSO looking for a single comma separated header for roles: https://docs.graylog.org/en/3.3/pages/users_and_roles/external_auth.html#single-sign-on

@stefansedich mentioned he did some investigation and saw a split of `,` `;` and `|` with `,` being most common - hence revisiting the groups array separator might be valid.

**Comment by alanjcastonguay** [comment]
I am responsible for the environment @lilida mentioned earlier, now running https://github.com/kubernetes/ingress-nginx/releases/tag/controller-v0.35.0 with `nginx.ingress.kubernetes.io/auth-response-headers: "x-auth-request-user, x-auth-request-email, x-auth-request-groups, authorization"` and `nginx.ingress.kubernetes.io/auth-url` pointed to an `oauth2-proxy` built from @lilida 's `multivalueheaders` branch and `--upstream=file:///dev/null`. The way nginx collects auth-subrequest response headers and sends them on to a backend application is important.

I can see merits of both a `something, delimited, list` of groups in one header, or a header repeated multiple times. I have only seen the single-header-with-list style in the wild. It seems pragmatic to support, even if less clean.

I would like to get back to mainline. Go is not my first language. I'm not sure what I could contribute. I am happy to research or test things.

**Comment by NickMeves** [comment]
@JoelSpeed Circling back on this one for your final opinion.

I personally lean toward comma delimited instead of split headers.  Even though split headers might be more to spec, the more prevalent implementation in upstream Commercial-off-the-Shelf software is to pull a comma delimited list of groups from a single header.

We also could add another config option to flip which type of group headers we support... But I hate the configuration explosion for simple things like this and would lean toward us picking a route.

**Comment by alanjcastonguay** [comment]
FWIW, I would want CSV as the primary mode, with split-headers or alternate-delimiters options added later if/when demand appears.

**Comment by NickMeves** [comment]
Something we debated in the original PR that looks like it will be an issue for some users afterall -- whether to allow configuring the group separator in the headers.  This question came up in Slack and this sample groups-like claim would have issues with a `,` separator by default:

```
  "memberof": [
        "CN=EMSUDEMY,OU=Groups,O=cco.xxx.com",
        "CN=all-xxxx-people,OU=xxxx Groups,O=cco.xxxx.com"
    ],
```

File: oauthproxy.go
Comment (NickMeves) [inline oauthproxy.go:949]: Do we want this as `req.Header["X-Forwarded-Groups"]` to match the convention set above in the other `X-Forwarded-*` request headers?

Either that or `Set` right? I don't think `Add` is correct here?

**Comment by alanjcastonguay** [comment]
Now that #839 is merged, can this move forward?

**Comment by NickMeves** [comment]
Can you take a peek at this PR: https://github.com/oauth2-proxy/oauth2-proxy/pull/826

When @JoelSpeed is back that is the next PR that is a baseline for a bunch of others we are looking to wrap up. It completely removes this codepath, but we are trying to work out the right design for `Get` vs `Set` that you have uncovered in this PR.

**Comment by JoelSpeed** [comment]
I'm wondering if we leave all of the existing logic as is, post #826, then maybe add some sort of flattening logic that reads the headers and flattens them as desired? Could become a new part of the header middlewares 🤔

**Comment by JoelSpeed** [comment]
#826 has now merged, I think the route forward is to have some logic that flattens headers in the header injection stuff, should be able to iterate through the headers, get the `Values()` (`[]string`), then do the `strings.Join` on that and finally `Set` the header to that value.

Do we want to support different separators for this or can we stick to comma separated in the first instance?

**Comment by therc** [comment]
Is this in scope for 7.0? We'd rather not run a fork to get the issue fixed.
Comma separated values might cover the vast majority of users for the initial release, it seems. No objection to that from here. :)

If the X.500-like Common Names are a concern, because multiple groups would be joined incorrectly into one line, breaking things even further, then one could check if the headers already contain a comma. If so, just return the first header as is, which is the same as the current behaviour for such environments, but unblocks the common case, no? Those users could then file an issue for custom separators — hoping that e.g. they don't require also some escaping/massaging of the values.

The logic above could be extended so that you go through e.g. the string `,;|` and use the first separator that is not present in the first header. There might be too much magic in that, though. For one, it considers the semicolon as the best choice if the incoming headers have a comma, which might or might not be a major assumption to make. (Also: do you repeat this for every request, or do you cache the inferred separator?)

**Comment by JoelSpeed** [comment]
This is a change in behaviour for us that could break users, I prefer we get this in before we release v7

**Comment by NickMeves** [comment]
> #826 has now merged, I think the route forward is to have some logic that flattens headers in the header injection stuff, should be able to iterate through the headers, get the `Values()` (`[]string`), then do the `strings.Join` on that and finally `Set` the header to that value.
> 
> Do we want to support different separators for this or can we stick to comma separated in the first instance?

Let's just support comma and see if anyone complains. At that point we can embark on another flag to set the array header separator.

I think in my research on this, comma-separated seemed to be the resounding standard for header consumers.

**Comment by JoelSpeed** [comment]
@lilida Do you have any time over the next couple of weeks to get this rebased and finished up? This is one of three PRs we are looking at for v7 which we are hoping to release before the end of the month. If you don't have time, do you mind if one of our maintainers finishes it up? (With appropriate attribution of course)

**Comment by lilida** [comment]
> @lilida Do you have any time over the next couple of weeks to get this rebased and finished up? This is one of three PRs we are looking at for v7 which we are hoping to release before the end of the month. If you don't have time, do you mind if one of our maintainers finishes it up? (With appropriate attribution of course)

Sure! I will get it finished

**Comment by lilida** [comment]
@JoelSpeed  I updated the pull request after rebase. I only update the claim injector and let me know if this is what you want

File: pkg/header/injector.go
Comment (JoelSpeed) [inline pkg/header/injector.go:62]: Would this be simpler if instead of making this add to the existing values, we just had the logic in a slightly different place and just flattened the values?

I was thinking we would have something like this in the middleware package that we could call after the header injection

```
func flattenHeaders(headers http.Header) {
  for name, values := range headers {
    headers.Set(name, strings.Join(values, ",")
  }
}
```

This could be called in the injectRequest and injectResponse middlewares after the headers have been injected
Requests:
https://github.com/oauth2-proxy/oauth2-proxy/blob/8e02fac2cc0ae61552d3244b313eb8e0dcd2d545/pkg/middleware/headers.go#L68
Response:
https://github.com/oauth2-proxy/oauth2-proxy/blob/8e02fac2cc0ae61552d3244b313eb8e0dcd2d545/pkg/middleware/headers.go#L99

**Review by JoelSpeed — COMMENTED** [review]
I think we can make this simpler, see my comment!

**Comment by JoelSpeed** [comment]
Great, I think this looks good now, could you add a changelog entry and make sure your branch is up to date with master please? (Rebase or merge, up to you)

@NickMeves Just wanted to check you agree this is good?

File: oauthproxy_test.go
Comment (NickMeves) [inline oauthproxy_test.go:584]: I don't think this is correct, the `groups` in the session will still be a list. This is cheating and is testing that a singleton group with the name `a,b` is reflected back as `a,b,` in the header.

We need to test that a list of groups here turns into a comma-delimited string in `X-Forwarded-Groups` in the final `assert`.

File: pkg/middleware/headers.go
Comment (NickMeves) [inline pkg/middleware/headers.go:44]: I think this is fine, right? I just always get uncomfortable mutating a map inside of a loop on itself. Might be lingering paranoia from my python days where that would result in a `RuntimeError`.

The only area I had concern was the underlying `Set` might mutate the `name` with its call to `textproto.CanonicalMIMEHeaderKey` in `textproto.MIMEHeader(h)` wrapped call.

Safest route is an initial `range` on the `http.Header` to get the headers & values to mutate. Then iterating over that list and modifying the passed in `http.Header`

**Review by NickMeves — CHANGES_REQUESTED** [review]
I think the test coverage that was modified in `oauthproxy_test.go` isn't quite right, it won't be testing what we intend to test.

Other than that looks good.

File: pkg/middleware/headers.go
Comment (lilida) [reply pkg/middleware/headers.go:44]: This should be safe in golang. The CanonicalMIMEHeaderKey was called for both add and set. So when we hit the code here, all names should be canonicalized. I updated condition to len(values) > 1 so we only need to update when necessary and it is safe because len(nil) is  0

**Review by NickMeves — APPROVED** [review]
LGTM! Thanks for your patience getting this over the finish line!

**Review by JoelSpeed — COMMENTED** [review]
Please add a changelog entry and a note in the breaking changes section as well explaining the change in behaviour, and then I'm happy to get this merged! Thanks @lilida

**Comment by NickMeves** [comment]
> Please add a changelog entry and a note in the breaking changes section as well explaining the change in behaviour, and then I'm happy to get this merged! Thanks @lilida

@JoelSpeed Do we have any breaking changes that need documenting? I think all the array based headers this affects are related to `X-Forwarded-Groups` which are also new in v7.

I don't think this breaks existing functionality in v6.1.1, right?

**Comment by JoelSpeed** [comment]
> I don't think this breaks existing functionality in v6.1.1, right?

Good point... It's been quite some time since the last release that it's completely blurred in my head 😅

In which case, never mind, just a regular changelog will be fine

**Comment by lilida** [comment]
> Please add a changelog entry and a note in the breaking changes section as well explaining the change in behaviour, and then I'm happy to get this merged! Thanks @lilida

Done!

File: CHANGELOG.md
Comment (JoelSpeed) [inline CHANGELOG.md:6]: I think we can just move this under `## Changes since v6.1.1` since we didn't have multiple value headers before v6.1.1, 
Please just add the following in the list for below, around line 70 at the moment
```
- [#799](https://github.com/oauth2-proxy/oauth2-proxy/pull/799) Use comma separated multiple values for header (@lilida)
```

**Review by JoelSpeed — APPROVED** [review]
LGTM, thanks @lilida

**Comment by gabrielrotaru** [comment]
Hello Guys,

Thanks a lot for your effort. I tried to test this relase v7.0.0 but this bug for me wasn't fixed. Maybe i'm doing something wrong.
I made the intergration with keycloak and rundeck. Keaycloak it's integrated with ldap and i mapped the groups.

Here is evaluation of the Token.
{
  "realm_access": {
    "roles": [
      "offline_access",
      "uma_authorization"
    ]
  },
  "resource_access": {
    "account": {
      "roles": [
        "manage-account",
        "manage-account-links",
        "view-profile"
      ]
    }
  },
  "scope": "openid email profile",
  "email_verified": true,
  "name": "Gabriel Stefan",
  "groups": [
    "Cust-Team-Test1",
    "Cust-Team-Test2",
    "Cust-Team-Test3",
    "Cust-Team-Test4"
  ],
  "preferred_username": "qqqq5555",
  "given_name": "Gabriel",
  "family_name": "Stefan",
  "email": "gabriel.stefan@test.com"
}

In Oauth in the log:
127.0.0.1:54820 - gabriel.stefan@test.com [2021/01/28 12:48:07] [AuthSuccess] Authenticated via OAuth2: Session{email:gabriel.stefan@test.com user:35dabf13-2641-4489-8740-417eab0a67b9 PreferredUsername:qqqq5555 token:true id_token:true created:2021-01-28 12:48:07.151380145 +0100 CET m=+91.954623590 expires:2021-01-28 12:53:07.144340139 +0100 CET m=+391.947583622 refresh_token:true groups:[Cust-Team-Test1 Cust-Team-Test2 Cust-Team-Test3 Cust-Team-Test4]}

Capture: 
Frame 18: 585 bytes on wire (4680 bits), 585 bytes captured (4680 bits)
Ethernet II, Src: 00:00:00_00:00:00 (00:00:00:00:00:00), Dst: 00:00:00_00:00:00 (00:00:00:00:00:00)
Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
Transmission Control Protocol, Src Port: 3000, Dst Port: 54396, Seq: 1, Ack: 4951, Len: 531
Hypertext Transfer Protocol
    HTTP/1.0 202 Accepted\r\n
    X-Auth-Request-Email: gabriel.stefan@test.com\r\n
    X-Auth-Request-Groups: Cust-Team-Test1\r\n
    X-Auth-Request-Groups: Cust-Team-Test2\r\n
    X-Auth-Request-Groups: Cust-Team-Test3\r\n
    X-Auth-Request-Groups: Cust-Team-Test4\r\n
    X-Auth-Request-Preferred-Username: qqqq5555\r\n
    X-Auth-Request-User: 35dabf13-2641-4489-8740-417eab0a67b9\r\n
    Date: Thu, 28 Jan 2021 11:30:50 GMT\r\n
    Content-Length: 0\r\n
    \r\n
    [HTTP response 1/1]
    [Time since request: 0.000689000 seconds]
    [Request in frame: 16]


Endpoint at rundeck server:
Frame 26: 5091 bytes on wire (40728 bits), 5091 bytes captured (40728 bits)
Ethernet II, Src: 00:00:00_00:00:00 (00:00:00:00:00:00), Dst: 00:00:00_00:00:00 (00:00:00:00:00:00)
Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
Transmission Control Protocol, Src Port: 61456, Dst Port: 4440, Seq: 1, Ack: 1, Len: 5037
Hypertext Transfer Protocol
    GET / HTTP/1.0\r\n
    Host: rundeck-cloud.local\r\n
    X-Forwarded-Proto: https\r\n
    X-Forwarded-User: qqqq5555\r\n
    X-Forwarded-Roles: Cust-Team-Test1\r\n
    X-Email: gabriel.stefan@test.com\r\n
    Connection: close\r\n
    User-Agent: Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:84.0) Gecko/20100101 Firefox/84.0\r\n
    Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8\r\n
    Accept-Language: en-US,en;q=0.5\r\n

Basically my expectation was to have X-Forwarded-Roles: Cust-Team-Test1,Cust-Team-Test2,Cust-Team-Test3,Cust-Team-Test4

Nginx  Version:  nginx/1.18.0
Part of nginx conf:
                        auth_request_set $user   $upstream_http_x_auth_request_preferred_username;
			auth_request_set $email  $upstream_http_x_auth_request_email;
                        auth_request_set $groups  $upstream_http_x_auth_request_groups;
			auth_request_set $auth_cookie $upstream_http_set_cookie;
			proxy_set_header X-Forwarded-User  $user;
                        proxy_set_header X-Forwarded-Roles  $groups;
			proxy_set_header X-Email $email;

BR,
Gabriel

**Comment by NickMeves** [comment]
@gabrielrotaru Can you move this to a distinct bug issue so we can track the conversation now that this original PR has already merged?

**Comment by NickMeves** [comment]
Once something merges, these threads are out of sight, out of mind (at least as is relates to my emails & notification workflow so I can promptly respond to new issues moving forward) 😅

**Pull request merged**
